tests/unit/mathematics/linalg/operations/Eigen3_operations_unittest.cc

```diff
@@ -275,1817 +314,2459 @@ TEST(LinalgBackendEigen, add_vector)
 		EXPECT_NEAR(A[i], result[i], 1e-15);
 }
 
-TEST(LinalgBackendEigen, SGVector_add_scalar)
+template <typename ST>
+void check_SGVector_add_scalar()
 {
 	const index_t n = 4;
-	float64_t s = -0.3;
+	ST s = -0.3;
 
-	SGVector<float64_t> a(n);
+	SGVector<ST> a(n);
 	for (index_t i = 0; i < (index_t)a.size(); ++i)
 		a[i] = i;
-	SGVector<float64_t> orig = a.clone();
+	SGVector<ST> orig = a.clone();
 
 	add_scalar(a, s);
 
 	for (index_t i = 0; i < (index_t)a.size(); ++i)
 		EXPECT_NEAR(a[i], orig[i] + s, 1e-15);
 }
 
-TEST(LinalgBackendEigen, SGMatrix_add_scalar)
+template <typename ST>
+void check_SGMatrix_add_scalar()
 {
 	const index_t r = 4, c = 3;
-	float64_t s = 0.4;
+	ST s = 0.4;
 
-	SGMatrix<float64_t> a(r, c);
+	SGMatrix<ST> a(r, c);
 	for (index_t i = 0; i < (index_t)a.size(); ++i)
 		a[i] = i;
-	SGMatrix<float64_t> orig = a.clone();
+	SGMatrix<ST> orig = a.clone();
 
 	add_scalar(a, s);
 
 	for (index_t i = 0; i < (index_t)a.size(); ++i)
 		EXPECT_NEAR(a[i], orig[i] + s, 1e-15);
 }
 
-TEST(LinalgBackendEigen, center_matrix)
+template <typename ST>
+void check_SGMatrix_center_matrix()
 {
 	const index_t n = 3;
-	float64_t data[] = {0.8192343,  0.13191962, 0.50888604,
-	                    0.16857468, 0.24107738, 0.89455301,
-	                    0.40657379, 0.07902286, 0.24319651};
-	float64_t result[] = {0.25587541,  -0.1173183, -0.13855711,
-	                      -0.34283925, 0.04378442, 0.29905482,
-	                      0.08696383,  0.07353387, -0.16049771};
+	ST data[] = {0.5, 0.3, 0.4,
+	             0.4, 0.5, 0.3,
+	             0.3, 0.4, 0.5};
+	ST result[] = {0.1, -0.1, 0.0,
+	               0.0, 0.1, -0.1,
+	               -0.1, 0.0, 0.1};
 
-	SGMatrix<float64_t> m(data, n, n, false);
+	SGMatrix<ST> m(data, n, n, false);
 
 	center_matrix(m);
 
 	for (index_t i = 0; i < (index_t)m.size(); ++i)
-		EXPECT_NEAR(m[i], result[i], 1e-8);
+		EXPECT_NEAR(m[i], result[i], 1e-7);
 }
 
-TEST(LinalgBackendEigen, SGMatrix_cholesky_llt_lower)
+template <typename ST>
+void check_SGMatrix_cholesky_llt_lower()
 {
 	const index_t size=2;
-	SGMatrix<float64_t> m(size, size);
+	SGMatrix<ST> m(size, size);
+	typedef Matrix<ST, Dynamic, Dynamic> Mxx; // need to adapt the Eigen::Matrix to ST type
 
 	m(0,0)=2.0;
 	m(0,1)=1.0;
 	m(1,0)=1.0;
 	m(1,1)=2.5;
 
 	//lower triangular cholesky decomposition
-	SGMatrix<float64_t> L = cholesky_factor(m);
+	SGMatrix<ST> L = cholesky_factor(m);
+
 
-	Map<MatrixXd> map_A(m.matrix,m.num_rows,m.num_cols);
-	Map<MatrixXd> map_L(L.matrix,L.num_rows,L.num_cols);
+	Map<Mxx> map_A(m.matrix,m.num_rows,m.num_cols);
+	Map<Mxx> map_L(L.matrix,L.num_rows,L.num_cols);
 	EXPECT_NEAR((map_A-map_L*map_L.transpose()).norm(),
-		0.0, 1E-15);
+		0.0, 1E-6); // need to change the treshold to work with floats
```

Comment:
as an idea, we could use a templated function `get_epsilon<T>` that looks up the appropriate epsilons?